File: Algorithms/pr4_radix_sort.java
//Alessandro DiMarco
//Project 4 Radix Sort
//CS 323-33
//10/30/15

import java.io.*;
import java.util.Scanner;

class Node {
	private String data;
	private Node next;

	public Node (String d) {
		data = d;
		next = null;
	}

	public String getData() { return data; }

	public Node getNext()  { return next; }

	public void setData(String s) { data = s; }

	public void setNext(Node n)  { next = n; }
}

class Stack {
	private Node top;

	public Stack() {
		top = null;
	}

	public void push(String d) {
		Node n = new Node(d);
		n.setNext(top);
		top = n;
	}

	public String peek() {
		if (top == null) {
			System.out.println("Stack is empty");
			return "";
		}

		return top.getData();
	}

	public String pop() {
		if (top == null) {
			System.out.println("Stack is empty");
			return "";
		}

		String t = top.getData();
		top = top.getNext();
		return t;
	}

	public boolean isEmpty() {
		return top == null;
	}

	public void printStack(PrintWriter writer) {
		Node current = top;
		while (current != null) {
			writer.print(current.getData() + " ");
			current = current.getNext();
		}
		writer.print("\n\n");
	}
}

class Queue {
	private Node head;
	private Node tail;

	public Queue() {
		head = new Node("-999");
		tail = head;
	}

	public void enqueue(String d) {
		Node n = new Node(d);

		tail.setNext(n);
		tail = n;
	}

	public String dequeue() { 
		if (isEmpty()) {
			System.out.println("Queue is empty");
			return "";
		}

		String temp = head.getNext().getData();
		if (head.getNext().getNext() == null) {
			tail = head;
			head.setNext(null);
		}
		else {
			head.setNext(head.getNext().getNext());
		}

		return temp;
	}


	public boolean isEmpty() {
		return head.getNext() == null;
	}

	public void printQueue(PrintWriter writer) {
		if (isEmpty()) {
			System.out.println("Queue is empty");
			return;
		}

		Node current = head;
		while (current != null) {
			writer.print(current.getData() + " ");
			current = current.getNext();
		}

		writer.print("\n");
	}
}

class HashTable {
	public Queue table[];

	public HashTable() {
		table = new Queue[10];
		for (int i = 0; i < 10; ++i)
			table[i] = new Queue();
	}

	public
		void setVal(int index, String value) {
		table[index].enqueue(value);
	}

	public void printHashTable (String tableNumber, PrintWriter writer) {
		writer.print(tableNumber);
		writer.print("\n----------------------------------------------------\n");

		for (int i = 0; i < 10; ++i) {
			writer.print(i + " ");
			table[i].printQueue(writer);
			writer.print("\n");
		}
	}
}

class RadixSort {
	private HashTable tableOne;		//currentTable = true
	private HashTable tableTwo;		//currentTable = false
	private boolean currentTable;	//boolean to switch tables
	private int  currentDigit;		//track digit place value
	private int  totalDigits;

	public RadixSort(int max) {
		tableOne = new HashTable();
		tableTwo = new HashTable();
		totalDigits  = max;
		currentDigit = max-1;
		currentTable = true;

	}

	public int hash(String d) {
		int num = Character.getNumericValue(d.charAt(currentDigit)); //convert string d into number
		return (num%10);
	}

	void sort(Stack stk, PrintWriter writer) {
		stk.printStack(writer);
		
		//move items from stack to tableOne
		while (!stk.isEmpty()) {
			String str = stk.pop();
			int length = str.length();

			if (length < totalDigits) {
				//pad numbers with leading zeros
				str = String.format("%" + totalDigits + "s", str).replace(' ', '0');
			}

		 	tableOne.setVal(hash(str), str);
		}

		tableOne.printHashTable("TableOne", writer);
		--currentDigit;

		//cycle through digits and move tables
		for (; currentDigit >= 0; --currentDigit) {
			if (currentTable) {

				//Move number from tableOne to tableTwo
				for (int i = 0; i < 10; ++i) {
					while (!tableOne.table[i].isEmpty()) {
						String str = tableOne.table[i].dequeue();
						tableTwo.setVal(hash(str), str);
					}
				}
				tableTwo.printHashTable("TableTwo", writer);
			}
			else {
				//Move number from tabletwo to tableOne
				for (int i = 0; i < 10; ++i) {
	 					while (!tableTwo.table[i].isEmpty()) {
	 						String str = tableTwo.table[i].dequeue();
	 						tableOne.setVal(hash(str), str);
	 					}
	 			}

				tableOne.printHashTable("TableOne", writer);
			}

			currentTable ^= true; //switch currentTable
		}
	}
}


class main {
	public static void main(String[] args) throws Exception {
		if (args.length < 2) {
			System.out.println("Missing Arguments");
			System.exit(1);
		}

		File in_file  = new File(args[0]);
		File out_file = new File(args[1]);
		Scanner readFile   = new Scanner(in_file);
		PrintWriter writer = new PrintWriter(out_file);

		//search for largest number and add numbers to stack
		int num, maxNum = -1;
		Stack stk  = new Stack();
		while (readFile.hasNext()) {
			stk.push(readFile.next());	//push number into stack
			num = stk.peek().length();	//find the length of the number
			if (num > maxNum)
				maxNum = num;
		}
		readFile.close();

		//sort numbers
		RadixSort rdx = new RadixSort(maxNum);
		rdx.sort(stk, writer);

		writer.close();
	}
}

File: Algorithms/pr4_radix_sort.cpp
//Alessandro DiMarco
//Project 4 Radix Sort
//CS 323-33
//10/24/15

#include <iostream>
#include <fstream>
#include <iomanip>
#include <string>
#include <sstream>
using namespace std;

ifstream in_file;
ofstream out_file;

class Node
{
  string data;
  Node* next;

public:
  Node (string d)
  {
    data = d;
    next = nullptr;
  }

  string getData() { return data; }

  Node* getNext()  { return next; }

  void setData(string s) { data = s; }

  void setNext(Node* n)  { next = n; }
};

class Stack
{
  Node* top;

public:
  Stack()
  {
    top = nullptr;
  }

  ~Stack()
  {
    Node* ptr = top;
    Node* temp;
    while (ptr != nullptr)
    {
      temp = ptr;
      ptr  = ptr->getNext();
      delete temp;
    }
  }

  void push(string d)
  {
    Node* n = new Node(d);
    n->setNext(top);
    top = n;
  }

  string pop()
  {
    if (top == nullptr)
    {
      cout << "Stack is empty";
      return "";
    }

    string t = top->getData();
    top = top->getNext();
    return t;
  }

  bool isEmpty()
  {
    return top == nullptr;
  }

  void printStack()
  {
    Node* current = top;
    while (current != nullptr)
    {
      out_file << current->getData() << " ";
      current   = current->getNext();
    }
    out_file << endl;
  }
};

class Queue
{
  Node* head;
  Node* tail;

public:
  Queue()
  {
    head = new Node("-999");
    tail = head;
  }

  void enqueue(string d)
  {
    Node* n = new Node(d);

    tail->setNext(n);
    tail = n;
  }

  string dequeue()
  {
    if (isEmpty())
    {
      out_file << "Queue is empty" << endl;
      return "";
    }

    string temp = head->getNext()->getData();
    if (head->getNext()->getNext() == nullptr)
    {
      tail = head;
      head->setNext(nullptr);
    }
    else
    {
      head->setNext(head->getNext()->getNext());
    }

    return temp;
  }


  bool isEmpty()
  {
    return head->getNext() == nullptr;
  }

  ~Queue()
  {
    Node* ptr = head;
    Node* temp;
    while (ptr != nullptr)
    {
      temp = ptr;
      ptr  = ptr->getNext();
      delete temp;
    }
  }

  void printQueue()
  {
    if (isEmpty())
    {
      out_file << "Queue is empty" << endl;
      return;
    }

    Node* current = head;
    while (current != nullptr)
    {
      out_file << current->getData() << " ";
      current = current->getNext();
    }

    out_file << endl;
  }
};

class HashTable
{
public:
  Queue table[10];

  HashTable() {}

  void setVal(int index, string value)
  {
    table[index].enqueue(value);
  }

  void printHashTable (string tableNumber)
  {
    out_file << tableNumber << endl;
    out_file << "----------------------------------------------------\n";

    for (int i = 0; i < 10; ++i)
    {
      out_file << i << " ";
      table[i].printQueue();
      out_file << endl;
    }
  }
};

//global stack
Stack stk;

class RadixSort
{
  HashTable tableOne;		//currentTable = true
  HashTable tableTwo;		//currentTable = false
  bool currentTable;		//boolean to switch tables
  int  currentDigit;		//track digit place value
  int  totalDigits;

public:
  RadixSort(int max)
  {
    totalDigits     = max;
    currentDigit = max-1;
    currentTable = true;

  }

  int hash(string d)
  {
    int num = (d[currentDigit]-'0');	//convert string d into number
    return num%10;
  }

  void move(HashTable& A, HashTable& B)
  {
    string str;
    for (int i = 0; i < 10; ++i)
    {
      while (!A.table[i].isEmpty())
      {
        //move string from table A to table B
        str = A.table[i].dequeue();
        B.setVal(hash(str), str);
      }
    }
  }

  void sort()
  {
    //move items from stack to tableOne
    while (!stk.isEmpty())
    {
      string str = stk.pop();
      int length = str.size();

      if (length < totalDigits)
      {
        //pad numbers with leading zeros
        stringstream ss;
        ss << std::setw (totalDigits) << setfill('0') << str;
        str = ss.str();
      }

      tableOne.setVal(hash(str), str);
    }
    tableOne.printHashTable("TableOne");
    --currentDigit;

    for (; currentDigit >= 0; --currentDigit)
    {
      if (currentTable)
      {
        //Move number from tableOne to tableTwo
        move(tableOne, tableTwo);
        tableTwo.printHashTable("TableTwo");
      }
      else
      {
        //Move number from tabletwo to tableOne
        move(tableTwo, tableOne);
        tableOne.printHashTable("TableOne");
      }

      currentTable ^= true; //switch currentTable
    }
  }
};

int main (int argc, char** argv)
{
  if (argc < 3) {
    cout << "Missing arguments\n";
    return -1;
  }

  in_file.open(argv[1]);
  out_file.open(argv[2]);

  // debugging
  // in_file.open("4_radix_sort_data.txt");
  // out_file.open("output1.txt");

  //find largest integer
  string incoming;
  int maxDigits = -1;
  while (in_file >> incoming)
  {
    stk.push(incoming);
    if ((int)incoming.size() > maxDigits)
    maxDigits = incoming.size();
  }

  //print stack
  out_file << "Stack:\n";
  stk.printStack();
  out_file << "\n";

  //sort numbers
  RadixSort rdx(maxDigits);
  rdx.sort();

  in_file.close();
  out_file.close();
}
